### ISSUE TITLE
Horn's Parallel Analysis to determine nfactors

### ISSUE THREAD

**Issue by maxheld83** [aiorazabala/qmethod#68]
I initially found the decision on how many factors to extract one of the hardest, and thought that some of the criteria (Scree plot) aren't that convincing.

Watts and Stenner (2012) mention another possibility: Horn's parallel analysis (PA).
This essentially draws many random (!) samples with similar parameters as your actual data and than extracts principal components (unrotated) from those random (!) data.
As principal components *can* in fact arise in random data (they often do), PA gives you an estimate of any given number of extracted factors *not* being a product of such randomness (at the 95% or 99% e.g.).

I found this a strict, but very convincing standard: one of the first tests in any q study may indeed be to make sure that the observed factors are not, in fact, random.

It was pretty easy to do this in R, because there is already a package for Horn's PA out there.

I am now wondering whether I should wrap this in a function for this package, maybe so that when calling qmethod, specifying `nfactors` becomes optional.

This would be a pretty trivial, thin wrapper, *and* increase dependencies which, I guess, would be a bad thing.

On the other hand, a lot of users might not know about this kind of possibility, and it may be helpful to further push this kind if statistical rigor in Q.

Also, it would be one less thing to specify for users (if they so choose).

Any thoughts?

**Comment by maxheld83** [comment]
here's some [discussion on the matter on the listserv in 2008](https://listserv.kent.edu/cgi-bin/wa.exe?A2=ind0804&L=Q-METHOD&P=R383&1=Q-METHOD&9=A&J=on&X=1DA27F77980D4D1535&d=No+Match%3BMatch%3BMatches&z=4):

> I think Steven Brown’s point underscores a critical question here – are q-method users interested in factors, or in meaningfully interesting groupings or categories of people? Historically, because of the use of a form of factor analysis, we have called these ‘factors’ but I suggest that for most of us it is less important that what we find can be defended as ‘a factor’ (statistically defined) than a grouping of people in terms of their ideas and opinions. I suspect that it would theoretically be possible to arrive at such categories by the kinds of coding and sorting procedures typical of ‘qualitative’ research. There would have to be clear rules re making the categories (perhaps such as are implicit for most of us in the correlations etc. ‘internal’ to factor analysis). This would be a  terribly laborious procedure, but feasible. If this is how q-method had been developed, we would probably not talk about ‘factors’ at all, but about meaningful categories.
 
Stephen
 
From: Q Methodology Network [mailto:Q-METHOD@LISTSERV.KENT.EDU] On Behalf Of BROWN, STEVEN
Sent: 01 April 2008 13:24
To: Q-METHOD@LISTSERV.KENT.EDU
Subject: Re: On the permissiveness of the centroid method
 
But logical procedures such as Horn’s (1965), to whose article the link below is connected, are of little practical use if the factor analyst is pursuing a hunch or other abductory hypothesis, which itself may determine the number and character of the factors extracted, rather than the other way around.  It is one thing to use the scree test or some other algorithm (including Horn’s) to determine the number of factors (as a function of the surface features of the data) and then to declare this to be the objective reality that is now to be interpreted; it is another thing to be puzzled or curious about something and then to use the machinery of factor analysis as a way to probe the social or experimental space.  At the recent workshop in Bangkok, a number of participants gave their Q-sort renditions of what, in their judgment, the Thai government ought to do about the problem of avian influenza.  One factor solution relied upon so-called objective procedures (such as eigenvalues > 1.00 plus varimax rotation), but an alternative solution began with the abductory assumption that there ought to be a “medical” point of view that should be given special attention, and so the first factor was directed toward a group of public health participants.  Interestingly, a person trained in veterinary medicine was not on the medical factor and so provided the basis for a second factor (defined solely by the veterinarian).  Some other persons with no special knowledge about avian influenza were allowed to form a third factor, and several people were on none of these factors, but they were left undefined since it was clear that whatever their views were, they didn’t share factor space with the medical experts.  I doubt if Horn’s method would have left so many people undefined, but this illustrates how determining the number of factors can emerge from the investigative process rather than proceed it.  Incidentally, this isn’t to be interpreted as valorizing or giving special weight to the medical point of view; only to isolate it so that attention can focus on the view given by those with the most extensive experience with the subject matter, and who are apt to have disproportionate influence when it comes to making recommendations to the government.
 
Reference
Horn, J.L. (1965).  A rationale and test for the number of factors in factor analysis.  Psychometrika, 30, 179-185.
 
___________________________________________
*  _____  ______  ____  __ __  ____  ___ _  *  Steven R. Brown
| |  ___||_    _||  _ ||  |  ||  _ ||   | | |  Political Science
| |___  |  |  |  |  _| |  |  ||  _| |     | |  Kent State University
| |_____|  |__|  |____| \___/ |____||_|___| |  (sbrown@kent.edu)
*___________________________________________*_________________________
When I was a little kid, we had a quicksand box. I was an only child... eventually.

**Comment by aiorazabala** [comment]
Yeah, read that recently too in Watts & Stenner. An enhancement to keep in mind!

**Comment by maxheld83** [comment]
I'll try and get around to doing this sometime soon. Shouldn't be too hard, just a wrapper for one of several parallel analyses already implemented in R.

You think it would be valuable then to include this as a function wrapper within `qmethod`? 

Should we maybe have `qmethod` call this new function automatically when `nfactors=NULL`?

**Comment by aiorazabala** [comment]
Hummm, I'd say that we keep this separate, for the same reason as in #66

**Comment by maxheld83** [comment]
ok!
In that case there's really not much need to even add this to qmethod, correct?
Since it's trivial to use these packages ... not much to be gained by qmethod wrapper as I see it now.

Let's put this on backlog.

**Comment by aiorazabala** [comment]
Hey, this seems to perform parallel analysis:
http://www.statmethods.net/advstats/factor.html
Section *Determining the Number of Factors to Extract*
Function parallel() in pkg nFactors

**Comment by maxheld83** [comment]
this is using the special-purpose `paran()` package now, which is also written by one of the leading authors on the method (I consider that a pro, and it's also nice just as a matter of courtesy).

On the other hand, the `psych` package also has such a facility, and in terms of dependencies, it might be more convenient to use that one.

Will also again look at the nFactors package @aiorazabala suggested.

**Comment by maxheld83** [comment]
done.

**Comment by maxheld83** [comment]
I've stumbled across an interesting / potentially consequential problem / concern with parallel analysis.
With relatively *few* people-variables, a couple of studies I've looked at (mostly my own) *don't do so well*, under parallel analysis, meaning, quite a lot of factors will have to be dropped.
That seems ok, and maybe that's just the way it is with such small N data.
I *want* to stand by the rigorous criterion of parallel analysis.

However, I wonder how we can get out of this conundrum, and get to a place, where more factors can be retained –– which is desirable in Q, especially because a) we are not concerned with the relative frequency of occurence of factors and b) any given P-set is often assembled somewhat arbitrarily.
So we *want* to extract more factors, we just also want to be sure they're not random artefacts.

Since we *cannot* increase the number of item-cases (that would be the normal way to go about this problem in R-technique), because people can (for now) sort only so many items, the only way out would appear to be to add *people-variables*.
This *should* increase unadjusted eigenvalues, as similar people join the fray.

The question, is, however, whether by the same token, the discount factor necessitated by parallel analysis *also* increases, thereby negating any increase in unadjusted eigenvalues.

This seems complicated.

I've made some preliminary simulations in R about this issue, and also posted this on StackExchange: http://stats.stackexchange.com/questions/158311/how-do-random-data-eigenvalues-change-as-random-variables-are-added?noredirect=1#comment301632_158311

This is, for now, not strictly related to any R program functionality (but may one day #158).
I'm just posting it here for reference.
